cobalt/browser/metrics/cobalt_metrics_service_client.cc

```diff
@@ -0,0 +1,196 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cobalt/browser/metrics/cobalt_metrics_service_client.h"
+
+#include <stdint.h>
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "base/callback.h"
+#include "base/memory/singleton.h"
+#include "base/strings/string16.h"
+#include "base/time/time.h"
+#include "cobalt/browser/metrics/cobalt_enabled_state_provider.h"
+#include "cobalt/browser/metrics/cobalt_metrics_log_uploader.h"
+#include "cobalt/h5vcc/metric_event_handler_wrapper.h"
+#include "components/metrics/enabled_state_provider.h"
+#include "components/metrics/metrics_log_uploader.h"
+#include "components/metrics/metrics_pref_names.h"
+#include "components/metrics/metrics_reporting_default_state.h"
+#include "components/metrics/metrics_service.h"
+#include "components/metrics/metrics_service_client.h"
+#include "components/metrics/metrics_state_manager.h"
+#include "components/prefs/in_memory_pref_store.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+#include "components/prefs/pref_service_factory.h"
+#include "third_party/metrics_proto/system_profile.pb.h"
+
+namespace cobalt {
+namespace browser {
+namespace metrics {
+
+// How frequently to attempt uploading.
+const int kStandardUploadIntervalSeconds = 5 * 60;  // 5 minutes.
+
+void CobaltMetricsServiceClient::SetOnUploadHandler(
+    h5vcc::MetricEventHandlerWrapper* metric_event_handler) {
+  metric_event_handler_ = metric_event_handler;
+  if (log_uploader_) {
+    log_uploader_->SetOnUploadHandler(metric_event_handler_);
+  }
+}
+
+CobaltMetricsServiceClient::CobaltMetricsServiceClient(
+    ::metrics::MetricsStateManager* state_manager, PrefService* local_state)
+    : metrics_state_manager_(state_manager) {
+  metrics_service_ = std::make_unique<::metrics::MetricsService>(
+      metrics_state_manager_, this, local_state);
+}
+
+::metrics::MetricsService* CobaltMetricsServiceClient::GetMetricsService() {
+  return metrics_service_.get();
+}
+
+ukm::UkmService* CobaltMetricsServiceClient::GetUkmService() {
+  // TODO(b/284467142): UKM disabled in Cobalt currently, replace when UKM
+  // is supported.
+  return nullptr;
+}
+
+void CobaltMetricsServiceClient::SetMetricsClientId(
+    const std::string& client_id) {
+  // TODO(b/286066035): What to do with client id here?
+}
+
+int32_t CobaltMetricsServiceClient::GetProduct() {
+  // Note, Product is a Chrome concept and similar dimensions will get logged
+  // elsewhere downstream. This value doesn't matter.
+  return 0;
+}
+
+std::string CobaltMetricsServiceClient::GetApplicationLocale() {
+  // The locale will be populated by the web client, so return value is
+  // inconsequential.
+  return "en-US";
```

Comment:
We actually have Starboard-returned locale, which significantly affects client behavior. It's in technical requirements and determines initial language selection. Can we will this in here ?

Reply:
We could, but I think this fits into https://b.corp.google.com/issues/286884542. This field is really only a logging dimension and its one that is redundant. GEL will tack a locale code onto its logs before sending. I have strong doubts these fields will actually be useful outside of some type of rare edge cases (e.g., debugging locale mismatches in Cobalt/Kabuki).